Flaky Target Allocator test

There's a flaky test for the TA which is causing random failures for [unrelated PRs](https://github.com/open-telemetry/opentelemetry-operator/actions/runs/9462700135/job/26066267341?pr=3025)

ex: 
```
=== RUN   Test_runWatch/pod_delete
    collector_test.go:159: 
        	Error Trace:	/home/runner/work/opentelemetry-operator/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:161
        	            				/opt/hostedtoolcache/go/1.21.10/x64/src/runtime/asm_amd64.s:1650
        	Error:      	"map[test-pod1:test-pod1 test-pod2:test-pod2 test-pod3:test-pod3]" should have 1 item(s), but has 3
    collector_test.go:159: 
        	Error Trace:	/home/runner/work/opentelemetry-operator/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:162
        	            				/opt/hostedtoolcache/go/1.21.10/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Not equal: 
        	            	expected: map[string]*allocation.Collector{"test-pod1":(*allocation.Collector)(0xc000a0e480), "test-pod2":(*allocation.Collector)(0xc000a0e4b0), "test-pod3":(*allocation.Collector)(0xc000a0e4e0)}
        	            	actual  : map[string]*allocation.Collector{"test-pod1":(*allocation.Collector)(0xc00080b2c0)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,14 +1,4 @@
        	            	-(map[string]*allocation.Collector) (len=3) {
        	            	+(map[string]*allocation.Collector) (len=1) {
        	            	  (string) (len=9) "test-pod1": (*allocation.Collector)({
        	            	   Name: (string) (len=9) "test-pod1",
        	            	-  NodeName: (string) (len=9) "test-node",
        	            	-  NumTargets: (int) 0
        	            	- }),
        	            	- (string) (len=9) "test-pod2": (*allocation.Collector)({
        	            	-  Name: (string) (len=9) "test-pod2",
        	            	-  NodeName: (string) (len=9) "test-node",
        	            	-  NumTargets: (int) 0
        	            	- }),
        	            	- (string) (len=9) "test-pod3": (*allocation.Collector)({
        	            	-  Name: (string) (len=9) "test-pod3",
        	            	   NodeName: (string) (len=9) "test-node",
    collector_test.go:159: 
        	Error Trace:	/home/runner/work/opentelemetry-operator/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:159
        	Error:      	Condition never satisfied
        	Test:       	Test_runWatch/pod_delete
--- FAIL: Test_runWatch (1.01s)
```

Found a race con in the test:
```
/Users/jacob.aronoff/.asdf/installs/golang/1.21.4/go/bin/go tool test2json -t /private/var/folders/9k/rkfb5r2s2xq_3bv7l6jmr3pr0000gp/T/GoLand/___1Test_runWatch_in_github_com_open_telemetry_opentelemetry_operator_cmd_otel_allocator_collector.test -test.v -test.paniconexit0 -test.run ^\QTest_runWatch\E$
=== RUN   Test_runWatch
=== RUN   Test_runWatch/pod_add
    collector_test.go:159: 
        	Error Trace:	/Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:159
        	Error:      	Condition never satisfied
        	Test:       	Test_runWatch/pod_add
==================
WARNING: DATA RACE
Read at 0x00c0008ae080 by goroutine 60:
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.Test_runWatch.func3()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:166 +0x860
  testing.tRunner()
      /Users/jacob.aronoff/.asdf/installs/golang/1.19.1/packages/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.darwin-arm64/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /Users/jacob.aronoff/.asdf/installs/golang/1.19.1/packages/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.darwin-arm64/src/testing/testing.go:1742 +0x40

Previous write at 0x00c0008ae080 by goroutine 62:
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.Test_runWatch.func3.2.1()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:152 +0x80
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.(*Watcher).runOnCollectors()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector.go:139 +0x27c
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.(*Watcher).rateLimitedCollectorHandler()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector.go:120 +0x148
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.(*Watcher).Watch.gowrap1()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector.go:84 +0x64

Goroutine 60 (running) created at:
  testing.(*T).Run()
      /Users/jacob.aronoff/.asdf/installs/golang/1.19.1/packages/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.darwin-arm64/src/testing/testing.go:1742 +0x5e4
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.Test_runWatch()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:136 +0x870
  testing.tRunner()
      /Users/jacob.aronoff/.asdf/installs/golang/1.19.1/packages/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.darwin-arm64/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /Users/jacob.aronoff/.asdf/installs/golang/1.19.1/packages/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.darwin-arm64/src/testing/testing.go:1742 +0x40

Goroutine 62 (running) created at:
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.(*Watcher).Watch()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector.go:84 +0x30c
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.Test_runWatch.func3.2()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:149 +0x158
  github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/collector.Test_runWatch.func3.gowrap1()
      /Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:155 +0x9c
==================
    testing.go:1398: race detected during execution of test
=== RUN   Test_runWatch/pod_delete
    collector_test.go:159: 
        	Error Trace:	/Users/jacob.aronoff/workspace/playground/opentelemetry-operator/cmd/otel-allocator/collector/collector_test.go:159
        	Error:      	Condition never satisfied
        	Test:       	Test_runWatch/pod_delete
--- FAIL: Test_runWatch (0.00s)
    --- FAIL: Test_runWatch/pod_add (0.00s)

    --- FAIL: Test_runWatch/pod_delete (0.00s)


FAIL

```

I changed two things:
1. increased the wait time (I was able to repro this by shrinking it, I think our runners are probs just slow sometimes)
2. move the map access inside the eventually con that we have which is already taking a lock

--- a/cmd/otel-allocator/collector/collector_test.go
+++ b/cmd/otel-allocator/collector/collector_test.go
@@ -158,13 +158,11 @@ func Test_runWatch(t *testing.T) {
 
 			assert.EventuallyWithT(t, func(collect *assert.CollectT) {
 				mapMutex.Lock()
+				defer mapMutex.Unlock()
 				assert.Len(collect, actual, len(tt.want))
 				assert.Equal(collect, actual, tt.want)
-				defer mapMutex.Unlock()
-			}, time.Second, time.Millisecond)
-
-			// check if the metrics were emitted correctly
-			assert.Equal(t, testutil.ToFloat64(collectorsDiscovered), float64(len(actual)))
+				assert.Equal(collect, testutil.ToFloat64(collectorsDiscovered), float64(len(actual)))
+			}, time.Second*3, time.Millisecond)
 		})
 	}
 }

